lookout/style/typos/typos_correction/utils.py

```diff
@@ -0,0 +1,172 @@
+    with open(file, "r") as f:
+        for line in f:
+            split = line.split()
+            tokens.append(split[0])
+    return tokens
+
+
+def join_lists(lists) -> list:
+    return list(chain.from_iterable([l for l in lists]))
+
+
+def flatten(dataframe: pandas.DataFrame, column: str, new_column: str,
+            apply_function=lambda x: x) -> pandas.DataFrame:
+    """
+    Flatten dataframe on 'column' with extracted elements put to 'new_column'
+    """
+    other_columns = list(dataframe.columns)
+    flat_other = numpy.repeat(dataframe.loc[:, other_columns].values,
+                              repeats=numpy.array(dataframe[column].apply(
+                                  lambda x: len(apply_function(x))).tolist()),
+                              axis=0)
+    flat_column = list(itertools.chain.from_iterable(dataframe[column].apply(
+        apply_function).tolist()))
+    result = pandas.DataFrame(flat_other, columns=other_columns)
+    result[new_column] = flat_column
+    return result
+
+
+def flatten_data(data: pandas.DataFrame, new_column_name="identifier"):
+    apply_function = (lambda x: x) if isinstance(data.token_split[0], list) \
+        else (lambda x: str(x).split())
+    return flatten(data, 'token_split', new_column_name, apply_function=apply_function)
+
+
+def add_context_info(data: pandas.DataFrame) -> pandas.DataFrame:
+    if "before" in data.columns and "after" in data.columns:
+        return data
+
+    tokens = list(data.typo)
+
+    if "token_split" in data.columns:
+        token_split = list(data.token_split)
+        before = []
+        after = []
+
+        for row_number in range(len(data)):
+            if "token_split" in data.columns:
+                split = token_split[row_number]
+                if isinstance(split, str):
+                    split = split.split()
+                index = split.index(tokens[row_number])
+                before.append(split[:index])
+                after.append(split[index + 1:])
+
+    else:
+        before = [[] for _ in range(len(data))]
+        after = [[] for _ in range(len(data))]
+
+    data["before"] = before
+    data["after"] = after
+
+    return data
+
+
+def rank_candidates(candidates, pred_proba, n_candidates=None, return_all=True):
+    """
+    Suggest typos corrections base on candidates
+    and correctness probabilities.
+
+    candidates: dataframe with columns "id", "typo", "candidate"
+                and indexed by range(len(pred_proba)).
+
+    Returns suggestions: {id : [(candidate, correct_prob)]},
+                         candidates are sorted
+                         by correct_prob in a descending order.
+    """
+    suggestions = {}
+    corrections = []
+    for i in range(len(pred_proba)):
+        index = candidates.loc[i, "id"]
+        corrections.append((candidates.loc[i, "candidate"], pred_proba[i]))
+
+        if i < len(pred_proba) - 1 and candidates.loc[i + 1, "id"] == index:
+            continue
```

Comment:
this part looks like you flatten something (with `flatten_data` ?) function and then do something here. Maybe it is better not to flat at all?

It is hard to tell without a context.